cluster/kubernetes/namespacer.go

```diff
@@ -0,0 +1,56 @@
+package kubernetes
+
+import (
+	"fmt"
+
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached"
+
+	kresource "github.com/weaveworks/flux/cluster/kubernetes/resource"
+)
+
+type namespaceViaDiscovery struct {
+	fallbackNamespace string
+	disco             discovery.DiscoveryInterface
+}
+
+// NewNamespacer creates an implementation of Namespacer
+func NewNamespacer(k *Kubectl, d discovery.DiscoveryInterface) (*namespaceViaDiscovery, error) {
+	fallback, err := k.getDefaultNamespace()
+	if err != nil {
+		return nil, err
+	}
+	cachedDisco := cached.NewMemCacheClient(d)
+	// in client-go v9, the call of Invalidate is necessary to force
+	// it to query for initial data; in subsequent versions, it is a
+	// no-op reset of the cache validity, so safe to call.
+	cachedDisco.Invalidate()
```

Comment:
I am missing some sort of invalidation or watch of CRDTs. Is this coming later?

Reply:
Yeah, coming later, probably in the form of `NewNamespacerViaDiscovery` creating the watch.

Would you elaborate on the consolidation you'd expect to see? Do you mean, should the bit that loads resources from the cluster _also_ adjust namespaces?

Reply:
I mean using the same cache both for this and getting all the resources for garbage-collection.

Reply:
Oh good question -- I've made that tricky for myself, but it would be nice to be able to do it

Reply:
I don't think you can watch APIResources, which would be the ideal, but I think it's possible to watch custom resource definitions: https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/client/clientset/clientset/typed/apiextensions/v1beta1/customresourcedefinition.go

This is just a more responsive (or accurate) version of "invalidate on a ticker"; there would still be opportunity for getting namespacedness wrong if it were changed. TBD: what happens when it gets the namespacedness wrong...